tests/test_pricing_model.py

```diff
@@ -0,0 +1,1281 @@
+        self.share_price = share_price
+        self.init_share_price = init_share_price
+
+
+class TradeResult:
+    __test__ = False  # pytest: don't test this class
+
+    def __init__(self, without_fee_or_slippage, with_fee, without_fee, fee):
+        self.without_fee_or_slippage = without_fee_or_slippage
+        self.with_fee = with_fee
+        self.without_fee = without_fee
+        self.fee = fee
+
+    def compare_trade_results(self, other):
+        np.testing.assert_almost_equal(
+            self.without_fee_or_slippage, other.without_fee_or_slippage, err_msg="unexpected without_fee_or_slippage"
+        )
+        np.testing.assert_almost_equal(self.with_fee, other.with_fee, err_msg="unexpected without_fee")
+        np.testing.assert_almost_equal(self.without_fee, other.without_fee, err_msg="unexpected fee")
+        np.testing.assert_almost_equal(self.fee, other.fee, err_msg="unexpected with_fee")
+
+
+class TestHyperdrivePricingModel(unittest.TestCase):
+    def test_calc_in_given_out(self):
+        pricing_model = HyperdrivePricingModel(False)
+        test_cases = [
+            (  # test one, basic starting point
+                TestCaseCalcInGivenOut(
+                    out=100,  # how many tokens you expect to get
+                    share_reserves=100_000,  # base reserves (in share terms) base = share * share_price
+                    bond_reserves=100_000,  # PT reserves
+                    token_in="base",  # what token you're putting in
+                    fee_percent=0.1,  # fee percent (normally 10%)
+                    days_remaining=182.5,  # 6 months remaining
+                    time_stretch_apy=0.05,  # APY of 5% used to calculate time_stretch
+                    share_price=1,  # share price of the LP in the yield source
+                    init_share_price=1,  # original share price pool started
+                ),
+                # From the input, we have the following values:
+                # T = 22.1868770168519182502689135891
+                # τ = 0.0225358440315970471499308329778
+                # 1 - τ = 0.977464155968402952850069167022
+                # k = c/u*(u*z)**(1-τ) + (2*y + c*z)**(1-τ)
+                # k = 100000**0.9774641559684029528500691670222 + (2*100000 + 100000*1)**0.9774641559684029528500691670222
+                # k = 302929.51067963685
+                (
+                    TradeResult(
+                        # p = ((2y+cz)/uz)**τ
+                        #   = 1.0250671833648672
+                        # without_fee_or_slippage = 1/p * out = 97.55458141947516
+                        without_fee_or_slippage=97.55458141947516
+                        # fee is 10% of discount before slippage = (100-97.55601990513969)*0.1 = 2.4454185805248443*0.1 = 0.24454185805248443
+                        ,
+                        fee=0.24454185805248443
+                        # deltaZ = 1/u * (u/c*(k - (2*y + c*z - deltaY)**(1-τ)))**(1/(1-τ)) - z
```

Comment:
I think this is the equation for delta z prime, not delta z in the hackmd. Is that ok?

Reply:
Yes. @wakamex can you make this update when you update the comments for your test? I updated mine already.